# Mahalanobis Distance

Mahalanobis Distance is a multivariate measure of the distance between a point and a distribution. It's an effective way to detect outliers, especially when dealing with multidimensional data.

#### Steps to Calculate Mahalanobis Distance:

1. Calculate the covariance matrix of the dataset.
2. Calculate the inverse of the covariance matrix.
3. For each data point (x), calculate its Mahalanobis Distance using the formula: D² = (x - μ)ᵀ * S⁻¹ * (x - μ), where S⁻¹ is the inverse covariance matrix.

#### Mahalanobis Distance Formula

$$ D^2 = (x - \mu)^T S^{-1} (x - \mu) $$

where:
- $x$ is the data point.
- $\mu$ is the mean of the dataset.
- $S$ is the covariance matrix of the dataset.
- $S^{-1}$ is the inverse of the covariance matrix.

## Test Platform : Intel Mac on Bootcamp Win10

In [5]:
import numpy as np
import platform

print(f"Machine: {platform.machine()}")
print(f"NumPy version: {np.__version__}")
print("---")
np.show_config()

Machine: AMD64
NumPy version: 1.26.0
---
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: C:/Users/ds0/miniconda3/envs/ml/Library/include
    lib directory: C:/Users/ds0/miniconda3/envs/ml/Library/lib
    name: mkl-sdl
    pc file directory: C:\b\abs_9fu2cs2527\croot\numpy_and_numpy_base_1695830496596\_h_env\Library\lib\pkgconfig
    version: '2023.1'
  lapack:
    detection method: pkgconfig
    found: true
    include directory: C:/Users/ds0/miniconda3/envs/ml/Library/include
    lib directory: C:/Users/ds0/miniconda3/envs/ml/Library/lib
    name: mkl-sdl
    pc file directory: C:\b\abs_9fu2cs2527\croot\numpy_and_numpy_base_1695830496596\_h_env\Library\lib\pkgconfig
    version: '2023.1'
Compilers:
  c:
    commands: cl.exe
    linker: link
    name: msvc
    version: 19.29.30152
  c++:
    commands: cl.exe
    linker: link
    name: msvc
    version: 19.29.30152
  cython:
    commands: cython
    linker: cython
    name: cython
    v

In [6]:
from sklearn.neighbors import LocalOutlierFactor
import numpy as np
from scipy.stats import chi2


In [7]:
# Old Sample data
X = [[10, 10], [12, 12], [11, 11], [14, 14], [100, 100], [14, 14]]
clf = LocalOutlierFactor(n_neighbors=2)
outlier_labels = clf.fit_predict(X)
# -1 indicates an outlier
print(f"Outlier labels: {outlier_labels}")

Outlier labels: [ 1  1  1  1 -1  1]


In [8]:
# Outlier detection using Mahalanobis distance
from scipy.stats import chi2
X = np.array([[10, 10], [12, 12], [11, 11], [14, 14], [100, 100], [14, 14]])
# Calculate mean and covariance matrix

mean = np.mean(X, axis=0)
cov_matrix = np.cov(X.T)
inv_cov_matrix = np.linalg.inv(cov_matrix)
cov_matrix

LinAlgError: Singular matrix

In [9]:
inv_cov_matrix

NameError: name 'inv_cov_matrix' is not defined

In [6]:
# Calculate Mahalanobis Distance for each data point
mahalanobis_distances = np.array([np.dot(np.dot((x - mean), inv_cov_matrix), (x - mean).T) for x in X])
mahalanobis_distances

array([0.26302083, 0.11588542, 0.12369792, 0.10026042, 4.57291667,
       0.10026042])

In [7]:
# Compare with a critical value from the chi-squared distribution
critical_value = chi2.ppf((1-0.1), df=2)  # df is the number of dimensions
outliers = np.where(mahalanobis_distances > critical_value)
critical_value

4.605170185988092

In [8]:
print(f"Outlier indices: {outliers[0]}")

Outlier indices: []


**The critical value is 4.605 whereas distance is 4.5729**

## Testing with New Sample

In [9]:
X = np.array([
    [10.0, 10.2], # Normal 1
    [10.2, 9.8],  # Normal 2
    [9.9, 10.1],  # Normal 3
    [10.1, 10.0], # Normal 4
    [10.0, 9.9],  # Normal 5
    [14.0, 18.0], # Outlier A (Caught by 0.90 threshold)
    [25.0, 5.0]   # Outlier B (Caught by 0.90 threshold)
], dtype=float)

In [10]:
clf = LocalOutlierFactor(n_neighbors=2)
outlier_labels = clf.fit_predict(X)
# -1 indicates an outlier
print(f"Outlier labels: {outlier_labels}")

Outlier labels: [ 1  1  1  1  1 -1 -1]


In [11]:
# Calculate mean and covariance matrix

mean = np.mean(X, axis=0)
cov_matrix = np.cov(X.T)
inv_cov_matrix = np.linalg.inv(cov_matrix)
cov_matrix

array([[31.39952381, -8.54642857],
       [-8.54642857, 14.63571429]])

In [12]:
inv_cov_matrix

array([[0.03786603, 0.02211162],
       [0.02211162, 0.08123795]])

In [13]:
# Calculate Mahalanobis Distance for each data point
mahalanobis_distances = np.array([np.dot(np.dot((x - mean), inv_cov_matrix), (x - mean).T) for x in X])
mahalanobis_distances

array([0.31684574, 0.34762866, 0.35610558, 0.32949348, 0.37168776,
       5.13786643, 5.14037235])

In [14]:
# Compare with a critical value from the chi-squared distribution
critical_value = chi2.ppf((1-0.1), df=2)  # df is the number of dimensions
outliers = np.where(mahalanobis_distances > critical_value)
critical_value

4.605170185988092

In [15]:
print(f"Outlier indices: {outliers[0]}")

Outlier indices: [5 6]


**The critical value is still 4.605 but distance is further away 5.13 and 5.14**